sol/main.c

```diff
@@ -0,0 +1,507 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <rte_sched.h>
+
+#include "gatekeeper_gk.h"
+#include "gatekeeper_launch.h"
+#include "gatekeeper_sol.h"
+
+/*
+ * Gatekeeper request priority queue implementation.
+ *
+ * To implement the request PQ, we maintain a linked list of packets
+ * listed in order of highest priority to lowest priority. We keep an
+ * array where each index represents a priority, and each element of
+ * that array holds a reference to the last packet of that priority.
+ * This allows us to quickly insert new packets of any priority into
+ * the linked list and drop the packet of lowest priority if necessary.
+ */
+
+/* The request priority queue is allocated 5% of the link's bandwidth. */
+#define REQ_QUEUE_BW_RATE (.05)
+
+/* Token bucket rate approximation error. */
+#define GATEKEEPER_TB_RATE_CONFIG_ERR (1e-7)
+
+/*
+ * Scaling for cycles_per_byte calculation; chosen so that minimum rate
+ * is 480 bit/sec. Adapted for use from the DPDK scheduling library.
+ */
+#define GATEKEEPER_TIME_SHIFT (8)
+
+static void
+insert_new_priority_req(struct req_queue *req_queue, struct priority_req *pr,
+	uint8_t priority)
+{
+	uint8_t next;
+
+	/*
+	 * Can't have these cases, since this function only inserts
+	 * a request for a priority that does not currently exist.
+	 */
+	RTE_VERIFY(priority != req_queue->highest_priority);
+	RTE_VERIFY(req_queue->priorities[priority] == NULL);
+
+	req_queue->priorities[priority] = pr;
+
+	/* This is the first packet in the queue. */
+	if (req_queue->len == 0) {
+		list_add(&pr->list, &req_queue->head);
+		req_queue->highest_priority = priority;
+		req_queue->lowest_priority = priority;
+		return;
+	}
+
+	/* Not the first packet, but still insert at the head of the queue. */
+	if (priority > req_queue->highest_priority) {
+		list_add(&pr->list, &req_queue->head);
+		req_queue->highest_priority = priority;
+		return;
+	}
+
+	/* Insert in middle or end of queue. */
+	if (priority < req_queue->lowest_priority)
+		req_queue->lowest_priority = priority;
+	for (next = priority + 1; next < req_queue->highest_priority; next++) {
+		if (req_queue->priorities[next] != NULL) {
+			list_add(&pr->list, &req_queue->priorities[next]->list);
+			return;
+		}
+	}
+
+	rte_panic("sol: %s failed to insert a request of a new priority\n",
+		__func__);
+}
+
+static void
+drop_lowest_priority_pkt(struct sol_config *sol_conf,
+	struct req_queue *req_queue)
+{
+	struct priority_req *lowest_pr =
+		req_queue->priorities[req_queue->lowest_priority];
+	struct rte_mbuf *lowest_pr_pkt = lowest_pr->pkt;
+	struct priority_req *next_lowest_pr;
+
+	RTE_VERIFY(req_queue->len > 0);
+
+	if (unlikely(req_queue->len == 1)) {
+		req_queue->priorities[lowest_pr->priority] = NULL;
+		req_queue->highest_priority = 0;
+		req_queue->lowest_priority = GK_NUM_REQ_PRIORITIES;
+		goto drop;
+	}
+
+	next_lowest_pr = list_last_entry(&lowest_pr->list,
+		struct priority_req, list);
```

Comment:
It works, but we really mean `list_prev_entry(&lowest_pr, list)` here. Please add `list_next_entry()` and `list_prev_entry()` to `include/list.h`.